numba/np/arraymath.py

```diff
@@ -3625,26 +3625,71 @@ def diff_impl(a, n=1):
     return diff_impl
 
 
-@overload(np.array_equal)
-def np_array_equal(a, b):
-
-    if not (type_can_asarray(a) and type_can_asarray(b)):
-        raise TypingError('Both arguments to "array_equals" must be array-like')
+if numpy_version >= (1, 19):
+    @overload(np.array_equal)
+    def np_array_equal(a, b, equal_nan=False):
+
+        if not (type_can_asarray(a) and type_can_asarray(b)):
+            raise TypingError('Both arguments to "array_equals" '
+                              'must be array-like')
+
+        if not isinstance(equal_nan, types.Boolean):
+            raise TypingError('"equal_nan" must be a boolean')
+
+        accepted = (types.Boolean, types.Number)
+        if isinstance(a, accepted) and isinstance(b, accepted):
+            # special case
+            def impl(a, b, equal_nan=False):
+                if equal_nan:
+                    if a != a and b != b:
+                        return True
+                return a == b
+        else:
+            def impl(a, b, equal_nan=False):
+                a = np.asarray(a)
+                b = np.asarray(b)
+                if a.shape != b.shape:
+                    return False
+                if equal_nan:
+                    for _a, _b in zip(a.flat, b.flat):
+                        if _a != _b:
+                            # NaN != NaN
+                            if _a != _a and _b != _b:
+                                continue
+                            else:
+                                return False
+                else:
+                    for _a, _b in zip(a.flat, b.flat):
+                        if _a != _b:
+                            return False
+                return True
 
-    accepted = (types.Boolean, types.Number)
-    if isinstance(a, accepted) and isinstance(b, accepted):
-        # special case
-        def impl(a, b):
-            return a == b
-    else:
-        def impl(a, b):
-            a = np.asarray(a)
-            b = np.asarray(b)
-            if a.shape == b.shape:
-                return np.all(a == b)
-            return False
+        return impl
+else:
+    @overload(np.array_equal)
+    def np_array_equal(a, b):
+
+        if not (type_can_asarray(a) and type_can_asarray(b)):
+            raise TypingError('Both arguments to "array_equals" '
+                              'must be array-like')
+
+        accepted = (types.Boolean, types.Number)
+        if isinstance(a, accepted) and isinstance(b, accepted):
+            # special case
+            def impl(a, b):
+                return a == b
+        else:
+            def impl(a, b):
+                a = np.asarray(a)
+                b = np.asarray(b)
+                if a.shape != b.shape:
+                    return False
+                for _a, _b in zip(a.flat, b.flat):
+                    if _a != _b:
+                        return False
```

Comment:
You can use [@register_jittable](https://github.com/numba/numba/blob/61ec1fd0f69aeadece218dccf4c39ebc5c7dfbc4/numba/core/extending.py#L138) to avoid duplicating. The two versions can share this loop; example:

```python
@register_jitable
def _array_equal_simple(a, b):
    for _a, _b in zip(a.flat, b.flat):
        if _a != _b:
            return False
```
so this `impl` can just be:

```python
 def impl(a, b):
    a = np.asarray(a)
    b = np.asarray(b)
    if a.shape != b.shape:
        return False
    return _array_equal_simple(a, b)
```

Reply:
updated